File: packages/shared/src/__tests__/link-router.test.ts
import type { Database as SqlJsDatabase } from 'sql.js'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import { AccountRepository } from '../data/account-repository'
import { ServicePluginRepository } from '../data/service-plugin-repository'
import { LinkRouter } from '../domain/link-router'
import { ServicePluginRegistry } from '../domain/service-plugin-registry'
import { createTestDatabase } from './test-database'

describe('LinkRouter', () => {
  let db: SqlJsDatabase
  let accountRepo: AccountRepository
  let registry: ServicePluginRegistry
  let router: LinkRouter

  beforeEach(async () => {
    db = await createTestDatabase()
    accountRepo = new AccountRepository(db)
    const pluginRepo = new ServicePluginRepository(db)
    registry = new ServicePluginRegistry(pluginRepo)
    registry.initialize()
    router = new LinkRouter(registry, accountRepo)
  })

  afterEach(() => {
    db.close()
  })

  function insertAccount(id: string, serviceId: string) {
    accountRepo.insert({
      accountId: id,
      serviceId,
      displayName: null,
      avatarUrl: null,
      sortOrder: 0,
      createdAt: '2026-01-01T00:00:00Z',
    })
  }

  it('classifies external links (unknown domain)', () => {
    const result = router.classifyLink('https://example.com/page', 'x')
    expect(result).toEqual({ type: 'external', url: 'https://example.com/page' })
  })

  it('classifies invalid URLs as external', () => {
    const result = router.classifyLink('not-a-url', 'x')
    expect(result).toEqual({ type: 'external', url: 'not-a-url' })
  })

  it('classifies same-domain navigation', () => {
    const result = router.classifyLink('https://x.com/notifications', 'x')
    expect(result).toEqual({ type: 'same-domain' })
  })

  it('classifies same-domain with subdomain (mobile.x.com)', () => {
    const result = router.classifyLink('https://mobile.x.com/home', 'x')
    expect(result).toEqual({ type: 'same-domain' })
  })

  it('classifies internal link with single account', () => {
    insertAccount('acc-ig-1', 'instagram')

    const result = router.classifyLink('https://instagram.com/p/123', 'x')
    expect(result).toEqual({
      type: 'internal-single-account',
      url: 'https://instagram.com/p/123',
      accountId: 'acc-ig-1',
    })
  })

  it('classifies internal link with multiple accounts', () => {
    insertAccount('acc-ig-1', 'instagram')
    insertAccount('acc-ig-2', 'instagram')

    const result = router.classifyLink('https://instagram.com/p/123', 'x')
    expect(result.type).toBe('internal-multi-account')
    if (result.type === 'internal-multi-account') {
      expect(result.serviceId).toBe('instagram')
      expect(result.accountIds).toHaveLength(2)
    }
  })

  it('classifies internal link with no account (unlogged service)', () => {
    // No instagram accounts exist
    const result = router.classifyLink('https://instagram.com/p/123', 'x')
    expect(result).toEqual({
      type: 'internal-no-account',
      url: 'https://instagram.com/p/123',
      serviceId: 'instagram',
    })
  })

  it('handles twitter.com as alias for X', () => {
    insertAccount('acc-x-1', 'x')

    // Navigating from instagram to twitter.com
    const result = router.classifyLink('https://twitter.com/user', 'instagram')
    expect(result).toEqual({
      type: 'internal-single-account',
      url: 'https://twitter.com/user',
      accountId: 'acc-x-1',
    })
  })

  it('handles www subdomain for known services', () => {
    insertAccount('acc-yt-1', 'youtube')

    const result = router.classifyLink('https://www.youtube.com/watch?v=abc', 'x')
    expect(result).toEqual({
      type: 'internal-single-account',
      url: 'https://www.youtube.com/watch?v=abc',
      accountId: 'acc-yt-1',
    })
  })
})

File: shared/src/jvmTest/kotlin/net/brightroom/uniso/domain/link/LinkRouterTest.kt
package net.brightroom.uniso.domain.link

import net.brightroom.uniso.data.repository.AccountRepository
import net.brightroom.uniso.data.repository.SessionRepository
import net.brightroom.uniso.data.repository.createTestDatabase
import net.brightroom.uniso.domain.account.AccountManager
import net.brightroom.uniso.domain.plugin.ServicePluginRegistry
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs

class LinkRouterTest {
    private lateinit var linkRouter: LinkRouter
    private lateinit var accountManager: AccountManager
    private lateinit var servicePluginRegistry: ServicePluginRegistry

    @BeforeTest
    fun setup() {
        val database = createTestDatabase()
        servicePluginRegistry = ServicePluginRegistry(database)
        accountManager =
            AccountManager(
                accountRepository = AccountRepository(database),
                sessionRepository = SessionRepository(database),
            )
    }

    private fun createRouter(): LinkRouter =
        LinkRouter(
            servicePluginRegistry = servicePluginRegistry,
            accountManager = accountManager,
        )

    // LR-001: External link → ExternalLink
    @Test
    fun externalLinkClassifiedAsExternalLink() {
        val account = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://example.com/article", account.accountId)

        assertIs<LinkClassification.ExternalLink>(result)
        assertEquals("https://example.com/article", result.url)
    }

    // LR-002: Internal link, single account → InternalSingleAccount
    @Test
    fun internalLinkWithSingleAccountClassifiedAsInternalSingleAccount() {
        // Source account is Instagram, target is X
        val igAccount = accountManager.addAccount("instagram").getOrThrow()
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://x.com/user", igAccount.accountId)

        assertIs<LinkClassification.InternalSingleAccount>(result)
        assertEquals("https://x.com/user", result.url)
        assertEquals(xAccount.accountId, result.account.accountId)
    }

    // LR-003: Internal link, multiple accounts → InternalMultiAccount
    @Test
    fun internalLinkWithMultipleAccountsClassifiedAsInternalMultiAccount() {
        // Source account is X, target is Instagram with 2 accounts
        val xAccount = accountManager.addAccount("x").getOrThrow()
        accountManager.addAccount("instagram")
        accountManager.addAccount("instagram")
        val router = createRouter()

        val result = router.classifyLink("https://instagram.com/p/123", xAccount.accountId)

        assertIs<LinkClassification.InternalMultiAccount>(result)
        assertEquals("https://instagram.com/p/123", result.url)
        assertEquals("instagram", result.serviceId)
        assertEquals(2, result.accounts.size)
    }

    // LR-004: Internal link, no logged-in account → InternalNoAccount
    @Test
    fun internalLinkWithNoAccountClassifiedAsInternalNoAccount() {
        // Source is X, target is Instagram but no Instagram accounts
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://instagram.com/user", xAccount.accountId)

        assertIs<LinkClassification.InternalNoAccount>(result)
        assertEquals("https://instagram.com/user", result.url)
    }

    // LR-005: Same domain navigation → SameDomainNavigation
    @Test
    fun sameDomainNavigationClassifiedAsSameDomainNavigation() {
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://x.com/home", xAccount.accountId)

        assertIs<LinkClassification.SameDomainNavigation>(result)
        assertEquals("https://x.com/home", result.url)
    }

    // LR-006: Subdomain support
    @Test
    fun subdomainMatchedCorrectly() {
        val igAccount = accountManager.addAccount("instagram").getOrThrow()
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://mobile.x.com/user", igAccount.accountId)

        assertIs<LinkClassification.InternalSingleAccount>(result)
        assertEquals(xAccount.accountId, result.account.accountId)
    }

    @Test
    fun handleNavigationReturnsFalseForSameDomain() {
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val shouldCancel =
            router.handleNavigation(
                url = "https://x.com/home",
                sourceAccountId = xAccount.accountId,
                onExternalLink = {},
                onSwitchAccount = { _, _ -> },
                onShowAccountSelector = {},
            )

        assertEquals(false, shouldCancel)
    }

    @Test
    fun handleNavigationReturnsTrueForExternalLink() {
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()
        var openedUrl: String? = null

        val shouldCancel =
            router.handleNavigation(
                url = "https://example.com/article",
                sourceAccountId = xAccount.accountId,
                onExternalLink = { openedUrl = it },
                onSwitchAccount = { _, _ -> },
                onShowAccountSelector = {},
            )

        assertEquals(true, shouldCancel)
        assertEquals("https://example.com/article", openedUrl)
    }

    @Test
    fun handleNavigationSwitchesAccountForSingleMatch() {
        val igAccount = accountManager.addAccount("instagram").getOrThrow()
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()
        var switchedTo: String? = null

        val shouldCancel =
            router.handleNavigation(
                url = "https://x.com/user",
                sourceAccountId = igAccount.accountId,
                onExternalLink = {},
                onSwitchAccount = { accountId, _ -> switchedTo = accountId },
                onShowAccountSelector = {},
            )

        assertEquals(true, shouldCancel)
        assertEquals(xAccount.accountId, switchedTo)
    }

    @Test
    fun handleNavigationShowsSelectorForMultipleAccounts() {
        val xAccount = accountManager.addAccount("x").getOrThrow()
        accountManager.addAccount("instagram")
        accountManager.addAccount("instagram")
        val router = createRouter()
        var selectorShown = false

        val shouldCancel =
            router.handleNavigation(
                url = "https://instagram.com/p/123",
                sourceAccountId = xAccount.accountId,
                onExternalLink = {},
                onSwitchAccount = { _, _ -> },
                onShowAccountSelector = { selectorShown = true },
            )

        assertEquals(true, shouldCancel)
        assertEquals(true, selectorShown)
    }

    @Test
    fun extractDomainHandlesValidUrl() {
        assertEquals("x.com", LinkRouter.extractDomain("https://x.com/user"))
        assertEquals("mobile.x.com", LinkRouter.extractDomain("https://mobile.x.com/home"))
        assertEquals("example.com", LinkRouter.extractDomain("https://example.com/path?q=1"))
    }

    @Test
    fun extractDomainReturnsNullForInvalidUrl() {
        assertEquals(null, LinkRouter.extractDomain("not-a-url"))
        assertEquals(null, LinkRouter.extractDomain(""))
    }

    @Test
    fun sameDomainViaAlternateDomain() {
        // X account should recognize twitter.com as same domain
        val xAccount = accountManager.addAccount("x").getOrThrow()
        val router = createRouter()

        val result = router.classifyLink("https://twitter.com/home", xAccount.accountId)

        assertIs<LinkClassification.SameDomainNavigation>(result)
    }
}
